src/shogun/clustering/KMeans.cpp

```diff
@@ -67,113 +66,41 @@ CKMeans::~CKMeans()
 
 void CKMeans::set_initial_centers(SGMatrix<float64_t> centers)
 {
-	dimensions = ((CDenseFeatures<float64_t>*) distance->get_lhs())->get_num_features();
+ 	CDenseFeatures<float64_t>* lhs=((CDenseFeatures<float64_t>*) distance->get_lhs());
+	dimensions=lhs->get_num_features();
 	REQUIRE(centers.num_cols == k,
 			"Expected %d initial cluster centers, got %d", k, centers.num_cols);
 	REQUIRE(centers.num_rows == dimensions,
 			"Expected %d dimensionional cluster centers, got %d", dimensions, centers.num_rows);
 	mus_initial = centers;
+	SG_UNREF(lhs);
 }
 
-void CKMeans::set_random_centers(SGVector<float64_t> weights_set, SGVector<int32_t> ClList, int32_t XSize)
+void CKMeans::set_random_centers()
 {
+	mus.zero();
 	CDenseFeatures<float64_t>* lhs=
 		CDenseFeatures<float64_t>::obtain_from_generic(distance->get_lhs());
-
-	for (int32_t i=0; i<XSize; i++)
-	{
-		const int32_t Cl=CMath::random(0, k-1);
-		weights_set[Cl]+=1.0;
-		ClList[i]=Cl;
-
-		int32_t vlen=0;
-		bool vfree=false;
-		float64_t* vec=lhs->get_feature_vector(i, vlen, vfree);
-
-		for (int32_t j=0; j<dimensions; j++)
-			mus.matrix[Cl*dimensions+j] += vec[j];
-
-		lhs->free_feature_vector(vec, i, vfree);
-	}
-
-	SG_UNREF(lhs);
+	int32_t lhs_size=lhs->get_num_vectors();
+	
+	SGVector<int32_t> temp=SGVector<int32_t>(lhs_size);
+	SGVector<int32_t>::range_fill_vector(temp, lhs_size, 0);
+	CMath::permute(temp);
 
 	for (int32_t i=0; i<k; i++)
 	{
-		if (weights_set[i]!=0.0)
-		{
-			for (int32_t j=0; j<dimensions; j++)
-				mus.matrix[i*dimensions+j] /= weights_set[i];
-		}
-	}
-}
-
-void CKMeans::set_initial_centers(SGVector<float64_t> weights_set,
-				SGVector<int32_t> ClList, int32_t XSize)
-{
-	ASSERT(mus_initial.matrix);
-
-	/// set rhs to mus_start
-	CDenseFeatures<float64_t>* rhs_mus=new CDenseFeatures<float64_t>(0);
-	CFeatures* rhs_cache=distance->replace_rhs(rhs_mus);
-	rhs_mus->copy_feature_matrix(mus_initial);
-
-	SGVector<float64_t> dists=SGVector<float64_t>(k*XSize);
-	dists.zero();
-
-	for(int32_t idx=0;idx<XSize;idx++)
-	{
-		for(int32_t m=0;m<k;m++)
-			dists[k*idx+m] = distance->distance(idx,m);
-	}
-
-	for (int32_t i=0; i<XSize; i++)
-	{
-		float64_t mini=dists[i*k];
-		int32_t Cl = 0, j;
-
-		for (j=1; j<k; j++)
-		{
-			if (dists[i*k+j]<mini)
-			{
-				Cl=j;
-				mini=dists[i*k+j];
-			}
-		}
-		ClList[i]=Cl;
-	}
-
-	/* Compute the sum of all points belonging to a cluster
-	 * and count the points */
-	CDenseFeatures<float64_t>* lhs=
-			(CDenseFeatures<float64_t>*)distance->get_lhs();
-	for (int32_t i=0; i<XSize; i++)
-	{
-		const int32_t Cl = ClList[i];
-		weights_set[Cl]+=1.0;
-
+		const int32_t Cl=temp[i];
```

Comment:
if you start touching things, we dont use Capital variable names...Also please give this a name that documents itself, i.e. "class"